### ISSUE TITLE
Lost volume after bounced the node pod

### ISSUE THREAD

**Issue by akunszt** [kubernetes-sigs/aws-efs-csi-driver#462]
/kind bug

**What happened?**
I'm in the middle of evaluating `aws-efs-csi`. I simulated a node pod failure/update process when the pods are bounced. I launched a test pod mounted an EFS PVC which wrote constantly to the EFS FileSystem. When the `efs-csi-node` pod bounced the test pod lost connectivity to the EFS and it never restored. Also every command issued on the host which accessed the mount point itself - for example a simple `df` - stuck. It looked like a typical NFS lost connection situation.

I'm using dynamic provisioning so one FileSystem with many AccesPoints.

**What you expected to happen?**

The EFS remains accessible when the `efs-csi-node` bounces. We used `efs-provisioner` before which just mounted the NFS filesystem and then let the Kubernetes handle it. I saw that `aws-efs-csi` uses `127.0.0.1` as the server which explains why everything stops working when the `efs-csi-node` has any issues.

```
127.0.0.1:/ /var/lib/kubelet/pods/0a00884f-eb72-41f5-b9c9-e8ed7a943716/volumes/kubernetes.io~csi/pvc-88c79104-4202-4914-a259-f92243bafb47/mount nfs4 rw,relatime,vers=4.1,rsize=1048576,wsize=1048576,namlen=255,hard,noresvport,proto=tcp,port=20376,timeo=600,retrans=2,sec=sys,clientaddr=127.0.0.1,local_lock=none,addr=127.0.0.1 0 0
```

Is it possible to mount the EFS FileSystem directly to avoid this single point of failure on the local node?

**How to reproduce it?**

Create an EFS PVC and launch a pod - for example `alpine:3.13` or any other general would do. Mount the EFS into the pod.

This is just an example how to do it.
```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: akunszt-test-efs
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1Mi
  storageClassName: aws-efs
---
apiVersion: v1
kind: Pod
metadata:
  name: akunszt-test
spec:
  restartPolicy: Never
  containers:
    - name: shell
      image: alpine:3.13
      command:
        - /bin/sh
      args:
        - "-c"
        - "apk update; apk add bash; while true; do sleep 60; done"
      volumeMounts:
        - name: pvc
          mountPath: /pvc
          readOnly: false
  volumes:
    - name: pvc
      persistentVolumeClaim:
        claimName: akunszt-test-efs
```
Exec into the pod - I use `bash` for the shell but that shouldn't matter - and execute the following command to constantly write to the PVC.
```sh
while true; do date '+%Y-%m-%d %H:%M:%S'; touch /pvc/a; sleep 0.2; done
```
Let it running and delete the `aws-efs-csi` pod running on the same node. The DaemonSet will re-create it instantly - like it would happen during an update - but the test pod will stuck/freeze, it won't retry - I waited about 20 minutes but even an `ls -l /pvc` freezes.

**Anything else we need to know?**:
It's possible - I hope - that I missed something during the deployment. Currently bouncing the `aws-efs-csi` pod could render the node unusable - for example monitoring or any other thing which uses just a `df` command in a script etc.

Also it blocks any intentional updates of the node pods as well as it will slowly kills/freezes every workload using EFS as the roll-out advances. Scaling down every Deployments/StatefulSets just to update `aws-efs-csi` is not an option. From the user's point of view this was unnoticeable before when we used `efs-provisioner`.

So I really hope that it's just me. In this case please point me to the right documentation or if it doesn't exist yet it's worth to add this scenario.

We are using `aws-vpc-cni` and `kiam`. We had to disable the `hostNetwork: true` in the chart to make it work at all but the pods still have routable IP addresses - assigned from the local VPC range - and can reach the EFS FileSystem directly.

**Environment**
- Kubernetes version: 1.21.1
- Driver version: 1.2.1

**Comment by kbasv** [comment]
Can you double check if the driver version is `1.2.1` ?

Also, can you run the [log collector](https://github.com/kubernetes-sigs/aws-efs-csi-driver/tree/master/troubleshooting). I'm interested to see the watchdog and mount logs from `efs-utils` which runs on the pod.

**Comment by akunszt** [comment]
@kbasv Thanks for looking into this. I double checked and it's really version `1.2.1`. We pulled the image from `docker.io` and pushed into our internal Docker registry.

```json
{
  "driverVersion": "v1.2.1-dirty",
  "gitCommit": "8b899c65c92ddb2e8f609cf9769ec32c87033205",
  "buildDate": "2021-04-29T19:36:56Z",
  "efsClientSource": "k8s",
  "goVersion": "go1.13.4",
  "compiler": "gc",
  "platform": "linux/amd64"
}
```
I replayed the scenario and ran the `log_collector.py` before and after bouncing the `efs-csi-node` pod on the node. I had to modify the script a bit to match our environment and I removed sensitive data from the results like the AWS AccountID.

The `results-pre-efs-csi-node-4xvbr.tar.gz` was made before bouncing the pod while the `results-post-efs-csi-node-ttslt.tar.gz` was made after that.

[results-post-efs-csi-node-ttslt.tar.gz](https://github.com/kubernetes-sigs/aws-efs-csi-driver/files/6608506/results-post-efs-csi-node-ttslt.tar.gz)
[results-pre-efs-csi-node-4xvbr.tar.gz](https://github.com/kubernetes-sigs/aws-efs-csi-driver/files/6608507/results-pre-efs-csi-node-4xvbr.tar.gz)

**Comment by kbasv** [comment]
@akunszt 
There was an stunnel issue on [efs-utils](https://github.com/aws/efs-utils) which failed to re-establish the connection after an `efs-csi-node` pod was bounced. We fixed this by upgrading efs-utils to v1.31.1 and it was released in efs-csi-driver version v1.3+.  I tested your setup with the driver version 1.3 and your pod above continues to write after `efs-csi-node` pod recovers when bounced.

```
2021-06-21 20:33:20
2021-06-21 20:33:21
2021-06-21 20:33:21 // Bounced efs-csi-node pod here
2021-06-21 20:33:30
2021-06-21 20:33:31
2021-06-21 20:33:31
```

mount-watchdog.log
```
2021-06-21 20:33:29,138 - INFO - amazon-efs-mount-watchdog, version 1.31.1, is enabled and started
2021-06-21 20:33:29,144 - WARNING - TLS tunnel for fs-xxxxxxxx.var.lib.kubelet.pods.5e5c4832-54dd-4df1-b6e2-ee62052bb55f.volumes.kubernetes.io~csi.pvc-d1f77c66-0ed9-45f3-a1f5-bbebf99ca60f.mount.20151 is not running
2021-06-21 20:33:29,144 - INFO - Starting TLS tunnel: "/usr/bin/stunnel /var/run/efs/stunnel-config.fs-xxxxxxxx.var.lib.kubelet.pods.5e5c4832-54dd-4df1-b6e2-ee62052bb55f.volumes.kubernetes.io~csi.pvc-d1f77c66-0ed9-45f3-a1f5-bbebf99ca60f.mount.20151"
2021-06-21 20:33:29,146 - INFO - Started TLS tunnel, pid: 20
```

**Comment by akunszt** [comment]
@kbasv Thank you very much for your efforts. We will test it as soon as we can (end of quarter, everybody is quite busy). I think until that happens we can close this issue. If we face any other issues then I'll open a new issue and refer to this one.
